bindgen/callbacks.rs

```diff
@@ -205,4 +205,6 @@ pub struct FieldInfo<'a> {
     pub type_name: &'a str,
     /// The name of the field.
     pub field_name: &'a str,
+    /// The name of the type of the field.
+    pub field_type_name: Option<&'a str>,
```

Comment:
IT seems we should probably include c++ namespaces and so on here? So, use the same name as all the blocklists etc use.

Reply:
uhhh can I be counseled on how that should be done because I am not very familiar with this codebase, much less C++

Reply:
I also should note that I specifically want `None` instead of the `unnamed at file...` output that currently is given in certain other cases, because I think that output is pointless, even harmful: it requires me to filter-by-string for anonymous types, which is much more error-prone and dependent on the vagaries of bindgen's output instead of using a type encoding.

Reply:
@emilio do we include c++ namespaces for the type name as well? otherwise it would be a bit inconsistent.

Reply:
@pvdrz type_name currently uses `canonical_name`, so yes, I think so.

Reply:
Should be easy enough to test in the `bindgen-integration` crate?

Reply:
I don't really need the canonical name, though, because I don't need `ExprEvalStep__bindgen_ty_1`, because what's important to me is that it is an _anonymous_ structure, and I don't necessarily want to reason about anonymous structures by their _name_. Maybe their other attributes, but the real answer for their name is the empty string.

Reply:
Well, sure, but my point is, if you have:

```c++
namespace foo {
  class Bar { ... };
}
namesace baz {
  class Bar { ... };
}
```

You want to be able to differentiate between one and the other right? btw, it's unclear to me what this does for pointer fields, templates, and so on. I assume the answer is reasonable, but still it'd be worth having some tests.

You might still can make the anonymous / non-anonymous distinction (`if thing.name().is_some() { Some(thing.canonical_name()) } else { None }` or same using `map` or what not)

Reply:
My answer is that I don't really care about whatever the answer is to that because I don't use C++, not personally, nor in a work context, and in the one case where I've tried to use bindgen against it in anger, its insistence on qualifying namespaces was actually a disadvantage as the namespaces were effectively meaningless and I would have been better off if the Rust code did not know about them (they were used as a crude implementation of parameterized modules, instead of using something sensible for this facility, like e.g. template parameters).

Reply:
To be honest, I'd rather see this working in C and sort of working in C++ that not seeing it working at all. I think it wouldn't be unreasonable to take this as it is and see if someone else actually ends up using this and requiring the `canonical_name` as we could either add it as a new field or use it instead of the regular name if a flag is enabled.

@workingjubilee @emilio does that works for both of you?

Reply:
Sure.

And I'd be happy to do the `.map`-based solution, but my main point of confusion at this juncture is in this code:

https://github.com/rust-lang/rust-bindgen/blob/2fb25e3be6a6a8e9ba13a8d1c736d5d8180dab1e/bindgen/ir/item.rs#L40-L68

This suggests it's more complicated than just calling `canonical_name()`, so I feel, basically, not qualified to evaluate what "canonical name" means in bindgen because it conflates C++ namespaces with generated symbols.